### ISSUE TITLE
Pointers in uniformity analysis

### ISSUE THREAD

**Issue by alan-baker** [gpuweb/gpuweb#2741]
We think the uniformity analysis needs some updates to better account for pointers. Pointers have relatively limited functionality in WGSL currently, but they can be used for:
* assignment to most address spaces in a local manner (e.g. `let x = &y; *x = ...;`)
* passed as parameters for private, workgroup and function address spaces

The missing uniformity aspects have to do with function address space pointers since private and workgroup end up being non-uniform by default. The analysis is missing an lvalue rule for the dereference operator and overall, doesn't seem to handle function scope variable passed as pointers to helper functions.

Consider the following:
```
fn bar(p : ptr<i32, function>, c : bool) {
  if (c) {
    *p = 0;
  }
}

fn foo() {
  var x : i32 = 1;
  // ...
  bar(&x, cond);
  // ...
  // use x
}
```

What value node should the "use x" statement use for `x`? The analysis doesn't capture the value update that occurs in `bar`, so if `c` is a non-uniform value the `x` will become non-uniform.

I can see two possibilities here:
1. Be conservative and have a value node for each pointer parameter that gets an edge to MayBeNonUniform, or
2. Model pointer parameters like an extra return value and to have a more accurate analysis

CC @jrprice @RobinMorisset

**Comment by dneto0** [comment]
Number 2. makes most sense to me.  It's not fundamentally new; it reuses existing concepts to analyze returns.

**Comment by RobinMorisset** [comment]
Either one could work. One point to note about option 2 is that unless I am missing something it would make the analysis time for each function be O(function size * number of parameters of pointer type) instead of just O(function size).

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-04-19-minutes#pointers-in-uniformity-analysis--issue-2741">WGSL meeting minutes 2022-04-19</a></summary>



* AB: When there’s a pointer parameter, should analysis be extra-conservative and pollute the call graph, or go more coarse grain and treat it as a return-value.  Think there is no big issue in performance when analyzing.   We could go either way here.
* MM: Don’t think we have an opinion.
* KG: Default desire is to add it to the pool of outputs. Like other returns. Outputs are returns and any other pointer params.
* MM: One day there will likely be an extension to allow pointers-to-pointers. How would that fit with this?
* AB: Depends on where you allow those pointers-to-pointers to exist.  The reason this doesn’t apply to workgroup and private is that those are always non-uniform because they are module-scope.  The only thing that could be uniform is function-scope.  Feel this would stay relatively contained. If we want to do something more broadly with module-scope variables, then consider that later.
* JB: Confident that once we start rejecting programs, we’ll make many adjustments.  Think we should work quickly to an implementation.
* AB: Don’t know what direction that is.
* JB: Smallest change that gets us to an implementation.
* DN: Thought both options were equally complicated, so leaned toward the more precise option.
* KG: Agree.
* MM: To Jim’s point about iterating.  It’s unlikely those source programs that will be tripped up will have pointers in them, so likely won’t have pushback on this particular issue.
* KG: Propose we try the precise analysis, and see if it works.
* JP: If we’re going with option 2, should have feedback this week.


</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-04-26-minutes#pointers-in-uniformity-analysis--issue-2741">WGSL meeting minutes 2022-04-26</a></summary>



* Previously
    * JP: If we’re going with option 2, should have feedback this week.
* JP: Implemented and works as expected. Still has complexity increas with number of function parameters, but we have a limit on those too.
* **SPEC NEEDED**


</details>

**Comment by dneto0** [comment]
This was resolved via #3298
